src/analysis/refactor_extract_region.ml

```diff
@@ -0,0 +1,479 @@
+open Std
+
+exception Nothing_to_do
+exception Not_allowed_in_interface_file
+
+let () =
+  Location.register_error_of_exn (function
+    | Nothing_to_do -> Some (Location.error "Nothing to do")
+    | Not_allowed_in_interface_file ->
+      Some
+        (Location.error
+           "Expression extraction is only allowed in implementation file")
+    | _ -> None)
+
+module Fresh_name = struct
+  (* Generate a fresh name that does not already exist in given environment. *)
+  let gen_val_name basename env =
+    let rec loop n =
+      let guess = basename ^ Int.to_string n in
+      if Env.bound_value guess env then succ n |> loop else guess
+    in
+    loop 1
+end
+
+let clean_up_for_printing expr =
+  let mapper =
+    { Ast_mapper.default_mapper with
+      expr =
+        (fun mapper expr ->
+          match expr.pexp_desc with
+          | Pexp_construct
+              ( ident,
+                Some
+                  { pexp_desc =
+                      Pexp_tuple
+                        (_
+                        :: ({ pexp_desc =
+                                Pexp_constant
+                                  { pconst_desc = Pconst_string _; _ };
+                              _
+                            } as const)
+                        :: _);
+                    _
+                  } )
+            when Longident.head ident.txt = "CamlinternalFormatBasics" ->
+            (* We need to retransform format specification which has been desugared into string. *)
+            const
+          | Pexp_poly (expr, _) ->
+            (* We also have to remove poly extra that cause unexpected "!poly!"
+               to be printed in generated code. This happens when you try
+               to extract the body of a method. *)
+            expr
+          | _ -> Ast_mapper.default_mapper.expr mapper expr)
+    }
+  in
+  mapper.expr mapper expr |> Parsetree_utils.expr_remove_merlin_attributes
+
+module Gen = struct
```

Comment:
I expect these utilities to land in a separate module common to multiple refactoring in the future. We should start documenting precisely their behavior. (the last few ones don't have a comment).

Reply:
Yeah it's the plan :slightly_smiling_face: (i added a comment for the two functions that didn't have one.)